### PR TITLE
chore: Upgrade to ESlint 9 and forbid import cycles

### PR DESCRIPTION
**Pull request by jmrossy**
### Description

- Upgrade from eslint 8 -> 9
- Forbid import cycles and self imports
- Fix lint errors caught by v9 but not v8
- Include utils package in CI linting

**No semantic changes, just some shuffling of things around to avoid cycles**

Fixes https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/4898

https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-cycle.md
https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-self-import.md

### CODE REVIEW

File: typescript/sdk/src/ism/metadata/builder.ts [hunk 1]
```diff
@@ -132,37 +101,4 @@ export class BaseMetadataBuilder implements MetadataBuilder {
         throw new Error(`Unsupported ISM: ${ism}`);
     }
   }
-
-  static decode(
```
Comment (jmrossy) [inline typescript/sdk/src/ism/metadata/builder.ts:136]: This was very problematic and I haven't completely sorted it out so I've lowered the lint severity to `warn` in this folder. 

We shouldn't have a web of static class methods calling each other or cycles will be inevitable. Also it makes usage of inheritance impractical so why use static methods in the first place.

File: typescript/sdk/src/ism/metadata/builder.ts [hunk 2]
```diff
@@ -132,37 +101,4 @@ export class BaseMetadataBuilder implements MetadataBuilder {
         throw new Error(`Unsupported ISM: ${ism}`);
     }
   }
-
```
Comment (jmrossy) [inline typescript/sdk/src/ism/metadata/builder.ts:135]: This was very problematic and I haven't completely sorted it out so I've lowered the lint severity to warn in this folder.

We shouldn't have a web of static class methods calling each other or cycles will be inevitable. Also it makes usage of inheritance impractical so why use static methods in the first place.

File: eslint.config.mjs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      'no-extra-boolean-cast': ['error'],
+      'no-ex-assign': ['error'],
+      'no-constant-condition': ['off'],
+      'no-return-await': ['error'],
+
+      'no-restricted-imports': [
+        'error',
+        {
+          name: 'console',
+          message: "Please use a logger and/or the utils' package assert",
```
Comment (ltyu) [inline eslint.config.mjs:83]:
```suggestion
          message: 'Please use a logger and/or the utils' package assert',
```

**Review by xeno097 — APPROVED** [review]

**Review by ltyu — APPROVED** [review]
LGTM

**Pull request merged**